geomesa-fs/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/FileMetadata.scala

```diff
@@ -0,0 +1,149 @@
+/***********************************************************************
+ * Copyright (c) 2013-2017 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.fs.storage.common
+
+import java.util
+import java.util.Collections
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.locks.ReentrantLock
+
+import com.typesafe.config._
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.locationtech.geomesa.fs.storage.api.{Metadata, PartitionScheme}
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.opengis.feature.simple.SimpleFeatureType
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+
+// TODO GEOMESA-1913 Use atomic file writing for metadata
+class FileMetadata protected[FileMetadata] (fs: FileSystem,
+                                            path: Path,
+                                            sft: SimpleFeatureType,
+                                            partitionScheme: PartitionScheme,
+                                            encoding: String,
+                                            partitions: Map[String, List[String]],
```

Comment:
Are we ever going to be in situation where the number of internal nodes/paths will be larger than we care to have in memory?  (I.e., would it be better to have an iterator of paths?)